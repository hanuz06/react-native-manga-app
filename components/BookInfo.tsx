import React, { memo } from "react";
import {
  StyleSheet,
  View,
  Image,
  ScrollView,
  SafeAreaView,
} from "react-native";

import { Surface, Title, Divider, Text, useTheme } from "react-native-paper";
import color from "color";
import * as Linking from "expo-linking";

import moment from "moment";

import Constants from "expo-constants";

import { IBookDetails } from "../types";

import { IMAGE_URL } from "react-native-dotenv";

const BookInfo = ({
  id,
  author,
  image,
  description,
  released,
  last_chapter_date,
  chapters,
  url,
  categories,
  title,
}: IBookDetails) => {
  const theme = useTheme();

  const contentColor = color(theme.colors.text).alpha(0.8).rgb().string();
  const imageBorderColor = color(theme.colors.text).alpha(0.15).rgb().string();

  const handlePress = () => Linking.openURL(url);

  return (
    <SafeAreaView style={styles.safeAreaView}>
      <ScrollView>
        <Surface style={styles.container}>
          <View style={styles.topRow}>
            <View style={styles.imageContainer}>
              <Image
                source={{ uri: `${IMAGE_URL}/${image}` }}
                style={[
                  styles.image,
                  {
                    borderColor: imageBorderColor,
                  },
                ]}
              />
            </View>
            <View style={styles.bookInfoContainer}>
              <Title style={styles.title}>{title}</Title>
              <Divider />
              <View style={styles.authorReleaseChapterContainer}>
                <Title style={styles.authorReleaseChapterStyle}>Author:</Title>
                <Text style={[styles.content, { color: contentColor }]}>
                  {author}
                </Text>
              </View>
              <Divider />
              <View style={styles.authorReleaseChapterContainer}>
                <Title style={styles.authorReleaseChapterStyle}>
                  Release date:
                </Title>
                <Text style={[styles.content, { color: contentColor }]}>
                  {released}
                </Text>
              </View>
              <View style={styles.authorReleaseChapterContainer}>
                <Title style={styles.authorReleaseChapterStyle}>
                  Total chapters:
                </Title>
                <Text style={[styles.content, { color: contentColor }]}>
                  {chapters.length}
                </Text>
              </View>
            </View>
          </View>
          <Divider />
          <View style={styles.bottomRow}>
            <Title style={styles.title}>Description</Title>
            <Text style={[styles.content, { color: contentColor }]}>
              {description}
            </Text>
            <Title style={styles.title}>Last updated:</Title>
            <Text style={[styles.content, { color: contentColor }]}>
              {last_chapter_date
                ? moment.unix(last_chapter_date).format("DD MMM YYYY ")
                : "Sorry no data"}
            </Text>
            <Title style={styles.title}>Categories:</Title>
            <Text style={[styles.content, { color: contentColor }]}>
              {categories && categories.join(", ")}
            </Text>
            <Title
              style={{ ...styles.title }}
              // onPress={handlePress}
            >
              Link
            </Title>
            <Text
              style={[styles.content, { color: "blue" }]}
              onPress={handlePress}
            >
              {url}
            </Text>
          </View>
        </Surface>
      </ScrollView>
    </SafeAreaView>
  );
};

export default memo(BookInfo);

const styles = StyleSheet.create({
  safeAreaView: {
    flex: 1,
    maxWidth: 450,
  },
  topRow: {
    fontFamily: "dancingFont",
    flexDirection: "row",
    alignItems: "center",
    marginBottom: 20,
  },
  imageContainer: {
    width: "35%",
    justifyContent: "center",
    alignItems: "center",
  },
  image: {
    borderWidth: StyleSheet.hairlineWidth,
    width: "100%",
    height: 200,
  },
  bookInfoContainer: {
    width: "65%",
    padding: 15,
    alignItems: "flex-start",
  },
  authorReleaseChapterContainer: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  authorReleaseChapterStyle: {
    fontSize: 20,
    marginRight: 5,
  },
  container: {
    flex: 1,
    padding: 20,
  },
  bottomRow: {
    marginVertical: 15,
  },
  title: {
    marginVertical: 5,
    fontSize: 22,
  },
  content: {
    fontSize: 18,
    lineHeight: 26,
  },
});
