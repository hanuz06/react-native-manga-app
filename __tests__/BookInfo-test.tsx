import "react-native";
import React from "react";
import BookInfo from "../components/BookInfo";

import renderer from "react-test-renderer";

const Props: any = {
  id: "",
  author: "",
  categories: [],
  chapters: [],
  description: "",
  image: "",
  last_chapter_date: 0,
  released: 0,
  title: "",
  url: "",
  contentColor: "black",
  imageBorderColor: "grey",
  handlePress: () => {},
};

describe("<BookInfo />", () => {
  it("match to snapshot", () => {
    const BookInfoPage = renderer.create(<BookInfo {...Props} />).toJSON();
    expect(BookInfoPage).toMatchSnapshot();
  });
});
