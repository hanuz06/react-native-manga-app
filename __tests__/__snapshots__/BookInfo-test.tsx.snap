// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookInfo /> match to snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "maxWidth": 450,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#ffffff",
            "flex": 1,
            "padding": 20,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 3,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 4,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "fontFamily": "dancingFont",
              "marginBottom": 20,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "width": "35%",
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "",
                }
              }
              style={
                Array [
                  Object {
                    "borderWidth": 0.5,
                    "height": 200,
                    "width": "100%",
                  },
                  Object {
                    "borderColor": "grey",
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "padding": 15,
                "width": "65%",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "500",
                      "textAlign": "left",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 20,
                        "letterSpacing": 0.15,
                        "lineHeight": 30,
                        "marginVertical": 2,
                      },
                      Object {
                        "fontSize": 22,
                        "marginVertical": 5,
                      },
                    ],
                    Array [
                      Object {
                        "fontSize": 20,
                        "letterSpacing": 0.15,
                        "lineHeight": 30,
                        "marginVertical": 2,
                      },
                      Object {
                        "fontSize": 22,
                        "marginVertical": 5,
                      },
                    ],
                  ],
                ]
              }
            >
              
            </Text>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "rgba(0, 0, 0, 0.12)",
                    "height": 0.5,
                  },
                  undefined,
                  undefined,
                ]
              }
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.87)",
                        "fontFamily": "System",
                        "fontWeight": "500",
                        "textAlign": "left",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontSize": 20,
                          "letterSpacing": 0.15,
                          "lineHeight": 30,
                          "marginVertical": 2,
                        },
                        Object {
                          "fontSize": 20,
                          "marginRight": 5,
                        },
                      ],
                      Array [
                        Object {
                          "fontSize": 20,
                          "letterSpacing": 0.15,
                          "lineHeight": 30,
                          "marginVertical": 2,
                        },
                        Object {
                          "fontSize": 20,
                          "marginRight": 5,
                        },
                      ],
                    ],
                  ]
                }
              >
                Author:
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 18,
                        "lineHeight": 26,
                      },
                      Object {
                        "color": "black",
                      },
                    ],
                  ]
                }
              >
                
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "rgba(0, 0, 0, 0.12)",
                    "height": 0.5,
                  },
                  undefined,
                  undefined,
                ]
              }
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.87)",
                        "fontFamily": "System",
                        "fontWeight": "500",
                        "textAlign": "left",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontSize": 20,
                          "letterSpacing": 0.15,
                          "lineHeight": 30,
                          "marginVertical": 2,
                        },
                        Object {
                          "fontSize": 20,
                          "marginRight": 5,
                        },
                      ],
                      Array [
                        Object {
                          "fontSize": 20,
                          "letterSpacing": 0.15,
                          "lineHeight": 30,
                          "marginVertical": 2,
                        },
                        Object {
                          "fontSize": 20,
                          "marginRight": 5,
                        },
                      ],
                    ],
                  ]
                }
              >
                Release date:
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 18,
                        "lineHeight": 26,
                      },
                      Object {
                        "color": "black",
                      },
                    ],
                  ]
                }
              >
                0
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.87)",
                        "fontFamily": "System",
                        "fontWeight": "500",
                        "textAlign": "left",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontSize": 20,
                          "letterSpacing": 0.15,
                          "lineHeight": 30,
                          "marginVertical": 2,
                        },
                        Object {
                          "fontSize": 20,
                          "marginRight": 5,
                        },
                      ],
                      Array [
                        Object {
                          "fontSize": 20,
                          "letterSpacing": 0.15,
                          "lineHeight": 30,
                          "marginVertical": 2,
                        },
                        Object {
                          "fontSize": 20,
                          "marginRight": 5,
                        },
                      ],
                    ],
                  ]
                }
              >
                Total chapters:
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 18,
                        "lineHeight": 26,
                      },
                      Object {
                        "color": "black",
                      },
                    ],
                  ]
                }
              >
                0
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgba(0, 0, 0, 0.12)",
                "height": 0.5,
              },
              undefined,
              undefined,
            ]
          }
        />
        <View
          style={
            Object {
              "marginVertical": 15,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 22,
                      "marginVertical": 5,
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 22,
                      "marginVertical": 5,
                    },
                  ],
                ],
              ]
            }
          >
            Description
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontSize": 18,
                    "lineHeight": 26,
                  },
                  Object {
                    "color": "black",
                  },
                ],
              ]
            }
          >
            
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 22,
                      "marginVertical": 5,
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 22,
                      "marginVertical": 5,
                    },
                  ],
                ],
              ]
            }
          >
            Last updated:
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontSize": 18,
                    "lineHeight": 26,
                  },
                  Object {
                    "color": "black",
                  },
                ],
              ]
            }
          >
            Sorry no data
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 22,
                      "marginVertical": 5,
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 22,
                      "marginVertical": 5,
                    },
                  ],
                ],
              ]
            }
          >
            Categories:
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontSize": 18,
                    "lineHeight": 26,
                  },
                  Object {
                    "color": "black",
                  },
                ],
              ]
            }
          >
            
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                    "fontFamily": "System",
                    "fontWeight": "500",
                    "textAlign": "left",
                    "writingDirection": "ltr",
                  },
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 22,
                      "marginVertical": 5,
                    },
                  ],
                  Array [
                    Object {
                      "fontSize": 20,
                      "letterSpacing": 0.15,
                      "lineHeight": 30,
                      "marginVertical": 2,
                    },
                    Object {
                      "fontSize": 22,
                      "marginVertical": 5,
                    },
                  ],
                ],
              ]
            }
          >
            Link
          </Text>
          <Text
            onPress={[Function]}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontSize": 18,
                    "lineHeight": 26,
                  },
                  Object {
                    "color": "blue",
                    "textDecorationLine": "underline",
                  },
                ],
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
